File: problems/Rust/16. 3Sum Closest.rs
impl Solution {
    pub fn three_sum_closest(mut nums: Vec<i32>, target: i32) -> i32 {
        nums.sort_unstable();
        let mut res = nums[0] + nums[1] + nums[2];

        for i in 0..nums.len() - 2 {
            let mut l = i + 1;
            let mut r = nums.len() - 1;
            while l < r {
                let csum = nums[i] + nums[l] + nums[r];
                if (csum - target).abs() < (res - target).abs() {
                    res = csum;
                }

                if csum < target {
                    l += 1;
                } else if csum > target {
                    r -= 1;
                } else {
                    return csum;
                }
            }
        }
        res
    }
}

File: problems/Java/16. 3Sum Closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        
        int res = nums[0] + nums[1] + nums[2];

        for (int i = 0; i < nums.length - 2; i++) {
            int l = i + 1;
            int r = nums.length - 1;
            while (l < r) {
                int csum = nums[i] + nums[l] + nums[r];
                if (Math.abs(csum - target) < Math.abs(res - target)) {
                    res = csum;
                }
                
                if (csum < target) {
                    l++;
                } else if (csum > target) {
                    r--;
                } else {
                    return csum;
                }
            }
        }
        return res;
    }
}